app/models/service.rb

```diff
@@ -203,17 +203,45 @@ def modify_power_state_delay(opts)
   end
 
   def power_states_match?(action)
-    if power_states.uniq == map_power_states(action)
-      options[:power_status] = "#{action}_complete"
-      update_attributes(:options => options)
-      return true
+    if composite? && (power_states.uniq == map_composite_power_states(action))
+      return update_power_status(action)
+    elsif atomic? && (power_states[0] == POWER_STATE_MAP[action])
+      return update_power_status(action)
     end
     false
   end
 
-  def map_power_states(action)
+  def map_composite_power_states(action)
     action_name = "#{action}_action"
-    service_resources.map(&action_name.to_sym).uniq.map { |x| Service::ACTION_RESPONSE[x] }.map { |x| Service::POWER_STATE_MAP[x] }
+    service_actions = service_resources.map(&action_name.to_sym).uniq
+
+    # We need to account for all nil :start_action or :stop_action attributes
+    #   in that case only we assume 'Power On' for the :start_action
+    #   and 'Power Off' for the :stop_action
+    if service_actions.compact.empty?
+      action_index = Service::ACTION_RESPONSE.values.index(action)
+      mod_resources = service_actions.each_with_index do |sa, i|
+        sa.nil? ? service_actions[i] = Service::ACTION_RESPONSE.to_a[action_index][0] : sa
+      end
+    else
+      mod_resources = service_actions
+    end
+
+    # Map out the final power state we should have for the passed in action
+    mod_resources.map { |x| Service::ACTION_RESPONSE[x] }.map { |x| Service::POWER_STATE_MAP[x] }
+  end
+
+  def update_power_status(action)
+    options[:power_status] = "#{action}_complete"
+    update_attributes(:options => options)
+  end
+
+  def atomic?
+    service_template.service_type == 'atomic'
+  end
+
+  def composite?
+    service_template.service_type == 'composite'
```

Comment:
Does it make sense to `delegate` the `service_type`?  Then you can remove the `service_template` reference from `atomic?` and `composite?`.